crates/ntx-builder/src/transaction.rs

```diff
@@ -0,0 +1,256 @@
+        let TransactionCandidate { account, notes } = tx;
+
+        tracing::Span::current().set_attribute("account.id", account.id());
+        tracing::Span::current().set_attribute("notes.count", notes.len());
+
+        // Work-around for `TransactionExecutor` not being `Send`.
+        tokio::task::spawn_blocking(move || {
+            let rt = tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .expect("runtime should be built");
+
+            rt.block_on(async move {
+                let mut notes = notes
+                    .into_iter()
+                    .map(|note| InputNote::Unauthenticated { note: note.into() })
+                    .collect::<Vec<_>>();
+                // We shuffle the notes here to prevent having a failing note always in front.
+                notes.shuffle(&mut rand::rng());
+                let notes = InputNotes::new(notes).map_err(NtxError::InputNotes)?;
+
+                let data_store = NtxDataStore::new(account, self.genesis_header.clone());
+
+                self.filter_notes(&data_store, notes)
+                    .and_then(|notes| self.execute(&data_store, notes))
+                    .and_then(|tx| self.prove(tx))
+                    .and_then(|tx| self.submit(tx))
+                    .await
+            })
+        })
+        .map_err(NtxError::Panic)
+        .await
+        .flatten_result()
+        .inspect_err(|err| tracing::Span::current().set_error(err))
+    }
+
+    /// Returns a set of input notes which can be successfully executed against the network account.
+    ///
+    /// The returned set is guaranteed to be non-empty.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if
+    /// - execution fails unexpectedly
+    /// - no notes are viable
+    #[instrument(target = COMPONENT, name = "ntx.execute_transaction.filter_notes", skip_all, err)]
+    async fn filter_notes(
+        &self,
+        data_store: &NtxDataStore,
+        notes: InputNotes<InputNote>,
+    ) -> NtxResult<InputNotes<InputNote>> {
+        let executor = TransactionExecutor::new(data_store, None);
+        let checker = NoteConsumptionChecker::new(&executor);
+
+        let notes = match checker
+            .check_notes_consumability(
+                data_store.account.id(),
+                BlockNumber::GENESIS,
+                notes.clone(),
+                TransactionArgs::default(),
+                Arc::new(DefaultSourceManager::default()),
+            )
+            .await
+        {
+            Ok(NoteAccountExecution::Success) => notes,
+            Ok(NoteAccountExecution::Failure { successful_notes, error, .. }) => {
+                let notes = successful_notes
+                    .into_iter()
+                    .map(|id| notes.iter().find(|note| note.id() == id).unwrap())
+                    .cloned()
+                    .collect::<Vec<InputNote>>();
```

Comment:
Not for this PR: I wonder if `NoteConsumptionChecker` should return successful notes as `Vec<Note>` for convenience. cc @Fumuran.

Reply:
Please, this would be useful :)

I was actually thinking the return type could always be:

```rust
struct NoteCheckResult {
    pub successful: Vec<Note>,
    pub remainder: Vec<Note>,
    pub error: Option<Err>,
}
```

Though maybe one still needs the overall `Result` wrapper depending on the inner failure reasons..

But essentially I thought `Result::Err` would be the same as `successful.is_empty()` at the moment.

Reply:
Created an issue: https://github.com/0xMiden/miden-base/issues/1566